### PR TITLE
feat: dorq auto login

### PR DESCRIPTION
**Pull request by jamesclark-Zapata**
# The problem

Our users currently have to manually open their web browser, log in, then paste a token in their terminal.

# This PR's solution

- Open the login URL in the user's default browser
- Listen on a local web server to receive the token
- Times out and gives instructions if the user doesn't manage to log in after 60 seconds
- Prints instructions on manually logging in if we were unable to open a web browser (without timeout)

# Checklist

_Check that this PR satisfies the following items:_

- [x] Tests have been added for new features/changed behavior (if no new features have been added, check the box).
- [x] The [changelog file](CHANGELOG.md) has been updated with a user-readable description of the changes (if the change isn't visible to the user in any way, check the box).
- [x] The PR's title is prefixed with `<feat/fix/chore/internal/docs>[!]:`

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -54,6 +54,8 @@ install_requires =
     inquirer~=3.0
     tabulate
     pygments~=2.0
+    # For automatic login
+    aiohttp~=3.8
```
Comment (alexjuda) [inline setup.cfg:58]: **non-blocking,question**: I like this approach of annotating a dependency with the reason why it was added. Do you think we should do this for all of our deps?

File: src/orquestra/sdk/_base/cli/_dorq/_login/_login.py [hunk 2]
```diff
@@ -31,6 +32,10 @@ def __init__(
         # data sources
         self._config_repo: _repos.ConfigRepo = config_repo
         self._runtime_repo: _repos.RuntimeRepo = runtime_repo
+        self._token_repo: _repos.TokenRepo = token_repo or _repos.TokenRepo(
+            runtime_repo=self._runtime_repo,
+            presenter=self._login_presenter,
```
Comment (alexjuda) [inline src/orquestra/sdk/_base/cli/_dorq/_login/_login.py:37]: **thought**: repos are "data" layer. Mixing `TokenRepo` with `LoginPresenter` feels like a lack of separation of concerns. Fortunately `presenter` doesn't seem to be used!

File: src/orquestra/sdk/_base/cli/_dorq/_repos.py [hunk 3]
```diff
@@ -502,3 +507,49 @@ def get_workflow_def(self, module, name: str) -> sdk.WorkflowDef:
...
+    """A web server for automatically logging in a user"""
+
+    def __init__(
+        self,
+        runtime_repo: RuntimeRepo,
+        presenter: _presenters.LoginPresenter,
+        login_server: t.Optional[LoginServer] = None,
+    ):
+        self._runtime_repo = runtime_repo
+        self._presenter = presenter
```
Comment (alexjuda) [inline src/orquestra/sdk/_base/cli/_dorq/_repos.py:522]: **issue**: can't see this being used anywhere. Is it needed?

Marking as blocker for the sake of separation of concerns in the "data" vs "presentation" layers.

File: src/orquestra/sdk/_base/cli/_dorq/_ui/_errors.py [hunk 4]
```diff
@@ -89,10 +89,20 @@ def _(e: exceptions.WorkflowRunNotFinished) -> ResponseStatusCode:
 @pretty_print_exception.register
 def _(e: ConnectionError) -> ResponseStatusCode:
     _print_traceback(e)
-    click.echo("Unable to connect to Ray.")
+    click.echo(f"{e}")
```
Comment (alexjuda) [inline src/orquestra/sdk/_base/cli/_dorq/_ui/_errors.py:92]: **question,non-blocking**: what will see the user when the CLI can't connect to Ray?

File: tests/cli/dorq/test_login_server.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    site = web.TCPSite(runner, "localhost", 8080)
+    try:
+        await site.start()
+
+        async with aiohttp.ClientSession() as session:
+            async with session.get(f"http://localhost:8080{url}") as response:
+                pass
+    finally:
+        await site.stop()
+    return response
```
Comment (alexjuda) [inline tests/cli/dorq/test_login_server.py:81]: **praise**: nice!

File: src/orquestra/sdk/_base/cli/_dorq/_login/_login_server.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        app = web.Application()
+        app["cluster_url"] = cluster_url
+        app.add_routes([web.get("/state", self._state_handler)])
+        runner = web.AppRunner(app)
+        await runner.setup()
+        try:
+            site = web.TCPSite(runner, listen_host, 0)
+            await site.start()
+            self._login_url = self._runtime_repo.get_login_url(
+                cluster_url, is_ce, runner.addresses[0][1]
```
Comment (alexjuda) [inline src/orquestra/sdk/_base/cli/_dorq/_login/_login_server.py:42]: **praise**: I really like the fact that it doesn't hardcode the port in the callback URL!

**Review by alexjuda — CHANGES_REQUESTED** [review]
LGTM overall! Marking as blocked to resolve the login presenter comment.

File: setup.cfg (hunk 1 above)
Comment (jamesclark-Zapata) [reply setup.cfg:58]: We did it for a few of them, but the new CLI stuff was just "CLI utilities"

**Review by alexjuda — APPROVED** [review]

File: src/orquestra/sdk/_base/cli/_dorq/_ui/_errors.py (hunk 4 above)
Comment (jamesclark-Zapata) [reply src/orquestra/sdk/_base/cli/_dorq/_ui/_errors.py:92]: Whatever the exception Ray provided. I had changed this cause a ConnectionError could have come from the HTTP requests.

**Pull request merged**